File: SpekioToolbox/Ahorn/entities/laserTurret.jl
module SpekioToolboxLaserTurret

using ..Ahorn, Maple

@mapdef Entity "SpekioToolbox/LaserTurret" LaserTurret(x::Int, y::Int, cooldown::String="2.0", timeOffset::Number=0.0, chargeTimer::Number=1.4, followTimer::Number=0.9, startActivated::Bool=true, laserColor::String="Red", silent::Bool=false, toggleFlag::String="", waitForPlayer::Bool=true)

Ahorn.editingOrder(entity::LaserTurret) = String["x", "y", "cooldown", "timeOffset", "chargeTimer", "followTimer",  "laserColor",  "toggleFlag","startActivated","waitForPlayer","silent"]

const placements = Ahorn.PlacementDict(
    "Laser Turret (Spekio's Toolbox)" => Ahorn.EntityPlacement(
        LaserTurret,
        "point",
	    Dict{String, Any}()
    )
)

Ahorn.editingOptions(entity::LaserTurret) = Dict{String, Any}(
  "laserColor" => String["Red", "Blue", "Green","Gray","Purple","Yellow"]
)

function Ahorn.selection(entity::LaserTurret)
    x, y = Ahorn.position(entity)
    return  Ahorn.Rectangle(x - 8, y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::LaserTurret, room::Maple.Room)
	c = entity.laserColor
	spritePath = ""
	
	if entity.startActivated
		spritePath = "objects/SpekioToolbox/laserTurrets/laserTurret"* c *"/laserTurret_on"
	else
		spritePath = "objects/SpekioToolbox/laserTurrets/laserTurret"* c *"/laserTurret_off"
	end
		
	Ahorn.drawSprite(ctx, spritePath, 0,0)
end

end

File: SpekioToolbox/Loenn/entities/laserTurret.lua
local drawableSprite = require("structs.drawable_sprite")
local utils = require("utils")
local jautils = require("mods").requireFromPlugin("libraries.jautils")

local laserTurret = {}

laserTurret.name = "SpekioToolbox/LaserTurret"
laserTurret.depth = -13000
laserTurret.justification = {0.5, 0.5}
laserTurret.fieldInformation = {
	laserColor = {
        options = {"Red", "Blue", "Green","Gray","Purple","Yellow"},
        editable = false
	}
}

laserTurret.placements = {
    name = "LaserTurret",
    data = {
		startActivated = true,
		timeOffset = 0,
		cooldown = 2.0,
		silent = false,
		waitForPlayer = true,
		laserColor = "Red",
		toggleFlag = "",
		chargeTimer = "1.4",
		followTimer = "0.9"
    }
}

laserTurret.fieldOrder = {"x", "y", "cooldown", "timeOffset", "chargeTimer", "followTimer", "startActivated", "laserColor", "silent", "toggleFlag","waitForPlayer"}

function laserTurret.sprite(room, entity)
	local sprite
	local laserColor = entity.laserColor or "Red"

	if entity.startActivated then
		sprite = drawableSprite.fromTexture("objects/SpekioToolbox/laserTurrets/laserTurret".. laserColor .."/laserTurret_on", entity)
	else
			sprite = drawableSprite.fromTexture("objects/SpekioToolbox/laserTurrets/laserTurret".. laserColor .."/laserTurret_off", entity)
	end
	
	local sprites = jautils.getBorder(sprite, utils.getColor("000000"))
	table.insert(sprites, sprite)
	return sprites
end

return laserTurret